File: FutureVoice/Services/SessionSummarizer.swift
import Foundation

/// Turning a finished talk into its review material: the summary itself, and
/// everything the app DERIVES from it — vocabulary, expressions, verified
/// grammar evidence, drill cards, carryover credit, the learner profile.
///
/// This is one place because it runs from two moments that must produce
/// IDENTICAL results:
///   • End of a live call (`ConversationView.endSession`), and
///   • "Generate it again" from the talk's book, when that first attempt
///     failed (`ConversationDetailView`).
/// The second exists because the first can fail — network, credits, a
/// truncated response — and the talk is saved BEFORE the analysis runs. A
/// session stuck at `summary == nil` isn't half a book, it's no book at all:
/// no drills, no scorecard, and nothing folded into the next conversation's
/// prompt. Until this type existed, that state was permanent.
@MainActor
enum SessionSummarizer {

    struct Result {
        let summary: SessionSummary
        /// The session as persisted — summary attached, and `topic` filled in
        /// from the generated title when the talk had no picked topic.
        let session: Session
    }

    /// What the wrap-up screen is allowed to say it is doing.
    ///
    /// Every field is a REAL count, reported at the point in `summarize` where
    /// that work actually finished — a wrap-up that narrates steps it isn't
    /// taking is worse than a spinner, because the next one is disbelieved
    /// too. `nil` means "not done yet"; the numbers below are the same ones
    /// the summary sheet then shows.
    ///
    /// The shape is a growing struct rather than a stream of events so a view
    /// can render it with no state of its own, and a caller that ignores it
    /// costs nothing.
    struct Progress: Equatable {
        // ── While the model is writing. The summary is ONE call that takes
        // most of the wait, and it writes its sections in a fixed order, so
        // the section it has finished is a real, observable fact — see
        // `absorb(partial:)`. Without these the board sat on step one and
        // then completed all at once, which is a spinner with extra steps.
        /// The transcript has been read and titled.
        var readBack = false
        /// `phrases_used` — the lines worth saying differently.
        var wroteCorrections = false
        /// `suggested_drills` — the review cards.
        var wroteDrills = false
        /// `expressions_used`.
        var wroteExpressions = false
        /// `grammar_errors`.
        var wroteGrammar = false

        // ── Final counts, once the local pass has verified everything. Each
        // is the number the summary sheet then shows.
        /// Corrected lines the talk produced.
        var phrases: Int?
        /// Words + expressions the learner used that are new to their pool.
        var words: Int?
        /// Expressions the fluent self used that the learner can take.
        var offered: Int?
        /// Corrections that survived the "they really said this" check.
        var corrections: Int?
        /// Things they'd been studying that came out unprompted.
        var carryovers: Int?
        /// Review cards minted from this talk.
        var cards: Int?
        /// Everything above is final.
        var finished = false

        /// Read the model's half-written JSON for which sections are DONE.
        ///
        /// A section is finished when the key AFTER it has appeared — the
        /// cheapest check that can't be fooled by a value still being written,
        /// and it needs no incremental JSON parsing. Key order is the schema's
        /// order in `ConversationEngine.summarySystemPrompt`; if that order
        /// changes, this reports the wrong step (never a wrong NUMBER — every
        /// count below still comes from the finished payload).
        mutating func absorb(partial: String) {
            guard !partial.isEmpty else { return }
            readBack = readBack || partial.contains("\"phrases_used\"")
            wroteCorrections = wroteCorrections || partial.contains("\"new_patterns_detected\"")
            wroteDrills = wroteDrills || partial.contains("\"expressions_used\"")
            wroteExpressions = wroteExpressions || partial.contains("\"weak_vocab_areas\"")
            wroteGrammar = wroteGrammar || partial.contains("\"overall_note\"")
        }
    }

    /// Whether a saved talk still owes the learner its review material.
    /// Turn-count guard mirrors `endSession`: an empty talk has nothing to
    /// analyze and must not offer a button that can only fail.
    static func needsSummary(_ session: Session) -> Bool {
        session.summary == nil && session.turns.contains { $0.role == .user }
    }

    /// Analyze `session`, persist it with its summary, and fold the outcome
    /// into every store that learns from a talk.
    ///
    /// Throws only on the Gemini call — everything after it is local. The
    /// caller's already-saved session row is left untouched on a throw, so a
    /// failure is always retryable and never partial.
    static func summarize(session: Session, appState: AppState,
                          onProgress: ((Progress) -> Void)? = nil) async throws -> Result {
        let turns = session.turns
        let sessionId = session.id
        var progress = Progress()
        func report(_ edit: (inout Progress) -> Void) {
            edit(&progress)
            onProgress?(progress)
        }

        let systemP = ConversationEngine.summarySystemPrompt(
            targetLanguage: session.targetLanguage,
            nativeLanguage: appState.nativeLanguage,
            profile: appState.learnerProfile,
            knownAboutUser: appState.persona?.knownFacts ?? [],
            // What a talk is allowed to yield follows how much was said in it.
            expressionBudget: ConversationEngine.expressionBudget(
                fluentTurns: turns.filter { $0.role == .fluentSelf }.count)
        )
        let transcript = ConversationEngine.formatTranscript(turns)
        let metrics = ScorecardMetrics.compute(turns: turns)
        let userMessage = """
        transcript:
        \(transcript)

        metrics:
        \(metrics.promptJSON())
        """
        // Stable idempotency key: retrying — by re-tapping End, or from the
        // book days later — re-runs the SAME logical request without a
        // second charge, as long as the transcript hasn't grown.
        // STREAMED, unlike the other analysis calls (see CLAUDE.md), and not
        // for latency: nothing here can be used before the payload closes, and
        // `sendJSONStreamAccumulating` still returns the fully decoded object,
        // so the result is byte-for-byte what the buffered call produced. What
        // the stream buys is the only honest way to move the wrap-up screen
        // while the model writes — a call that takes most of the wait and
        // reports nothing leaves the board on step one until everything
        // finishes at once. A deploy without SSE degrades to buffering, and
        // then `onPartial` simply never fires.
        func request() async throws -> ClaudeSummaryPayload {
            try await GeminiClient.shared.sendJSONStreamAccumulating(
                system: systemP,
                messages: [GeminiClient.Message(role: .user, content: userMessage)],
                // The schema's worst case is big: up to 15 grammar_errors (quote +
                // correction + a NATIVE-language note each), 5 phrases_used, 3-4
                // drills, expressions, and a 6-field scorecard whose notes are
                // also native-language. On top of that gen-3 counts THINKING
                // tokens against this same ceiling. At 1400 a long B2 session ran
                // out mid-JSON and surfaced as "reply hit the token ceiling" on
                // End — the talk was already saved, but its whole review yield
                // was lost. The ceiling is not billed, only tokens actually
                // produced, so the headroom is free.
                // 4096 until 2026-08-19: the expression budget now scales with
                // the talk (up to 14 per side), and the failures that were
                // showing up on the long talks this ceiling has to cover —
                // 9, 11, 25 and 31 turns, all NSCocoaErrorDomain 4864, a JSON
                // the decoder couldn't read — are what running out of room
                // mid-object looks like when the model stops short of the
                // MAX_TOKENS flag. Free headroom, so take it.
                maxTokens: 8192,
                purpose: "summary",
                idempotencyKey: "summary:\(sessionId.uuidString):\(turns.count)",
                onPartial: { partial in report { $0.absorb(partial: partial) } }
            )
        }

        let payload: ClaudeSummaryPayload
        do {
            payload = try await request()
        } catch let error where error.isMalformedModelOutput {
            // The learner did nothing wrong and has nothing to fix — the model
            // wrote a shape we couldn't read. Ask once more before making the
            // end of a talk look like a failure. Free: `record_free_usage`
            // dedupes on the idempotency key, so the second attempt neither
            // charges nor spends a slot of the daily cap, and the wait is one
            // request on a screen that is already showing a wrap-up.
            Telemetry.log("talk_summary_retry", [
                "detail": error.decodeDetail ?? "unknown",
                "turns": String(turns.count),
            ])
            payload = try await request()
        }
        var computed = payload.toDomain()
        report { $0.phrases = computed.phrasesUsed.count }

        // Fold the user's spoken words into the long-term vocab pool; the
        // freshly-used words ride along on the summary so the wrap-up can
        // celebrate concrete progress. The pool accepts each text once, so
        // when a RESUMED talk is summarized again the fresh batch covers only
        // the new turns — merge with the previous summary's list instead of
        // overwriting it, or every resume erased "words you used first".
        let userTexts = turns.filter { $0.role == .user }.map { $0.transcript }
        let freshWords = VocabStore.shared.ingest(
            sessionId: sessionId, userTexts: userTexts)
        let priorWords = session.summary?.newWordsUsed ?? []
        computed.newWordsUsed = priorWords + freshWords.filter { !priorWords.contains($0) }

        // Keep only expressions that literally appear in the user's own
        // turns — the LLM occasionally paraphrases, and we never show or
        // store an expression they didn't actually say.
        //
        // Compared with punctuation and casing stripped, the SAME way the
        // grammar quotes below are checked. A raw substring match silently
        // dropped good expressions over a comma, an apostrophe form or a
        // capital letter, which is why a talk full of usable phrases could
        // end up with one odd survivor.
        let haystack = CarryoverDetector.normalized(userTexts.joined(separator: " "))
        let verifiedExpressions = computed.expressionsUsed.filter { phrase in
            let needle = CarryoverDetector.normalized(phrase)
            return !needle.isEmpty && haystack.contains(needle)
        }
        #if DEBUG
        let dropped = computed.expressionsUsed.count - verifiedExpressions.count
        if dropped > 0 {
            NSLog("EXPRCAPTURE dropped %d of %d as not-verbatim", dropped, computed.expressionsUsed.count)
        }
        #endif
        computed.expressionsUsed = verifiedExpressions
        report { $0.words = freshWords.count + verifiedExpressions.count }
        // A pre-per-key-tracking session being re-analyzed: what it already
        // counted is its previous summary's list — seed the store so those
        // don't count twice (no-op when per-key data exists).
        VocabStore.shared.notePriorExpressions(
            sessionId: sessionId, phrases: session.summary?.expressionsUsed ?? [])
        VocabStore.shared.ingestExpressions(
            sessionId: sessionId, phrases: verifiedExpressions)

        // The other direction: phrases the FLUENT SELF used. Same verbatim
        // guard against ITS turns, and cheaper to satisfy — that text is
        // model-written, so there is no transcriber between the phrase and
        // the check. Anything the learner already produced is dropped: it is
        // evidence, and it is in the list above.
        //
        // A resumed talk re-summarizes from turn one, so the previous take's
        // phrases are merged rather than replaced — the same rule
        // `newWordsUsed` follows, for the same reason.
        let fluentHaystack = CarryoverDetector.normalized(
            turns.filter { $0.role == .fluentSelf }.map(\.transcript).joined(separator: " "))
        let alreadyMine = Set(verifiedExpressions.map(CarryoverDetector.normalized))
        var seenOffered = Set<String>()
        let verifiedOffered = computed.expressionsOffered.filter { phrase in
            let needle = CarryoverDetector.normalized(phrase)
            guard !needle.isEmpty, !alreadyMine.contains(needle),
                  !haystack.contains(needle),          // they said it too — not new material
                  fluentHaystack.contains(needle) else { return false }
            return seenOffered.insert(needle).inserted
        }
        let priorOffered = session.summary?.expressionsOffered ?? []
        computed.expressionsOffered = priorOffered + verifiedOffered.filter {
            !priorOffered.contains($0)
        }
        report { $0.offered = computed.expressionsOffered.count }

        // Same guard for grammar evidence: a quote the user can't find in
        // their own words destroys trust in the whole list. Compare with
        // punctuation/casing stripped — STT and the LLM disagree on those
        // even when the words match.
        func normalized(_ s: String) -> String {
            s.lowercased()
                .components(separatedBy: CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "'")).inverted)
                .filter { !$0.isEmpty }
                .joined(separator: " ")
        }
        let normalizedHaystack = normalized(haystack)
        computed.grammarIssues = computed.grammarIssues.filter {
            let needle = normalized($0.quote)
            // A "fix" that only touches punctuation/casing (normalized forms
            // identical) is a transcription nitpick, not a spoken grammar
            // error — drop it.
            return !needle.isEmpty && normalizedHaystack.contains(needle)
                && needle != normalized($0.correction)
        }
        report { $0.corrections = computed.grammarIssues.count }

        // Did anything they'd been studying actually come out of their mouth?
        // Runs against the drill cards as they stood BEFORE this session's own
        // corrections are ingested below, so a card minted tonight can't be
        // credited as carried into tonight.
        let carryovers = CarryoverDetector.detect(
            in: turns, cards: DrillStore.shared.load(),
            curriculumItems: appState.openCurriculumItems,
            studyingExpressions: VocabStore.shared.studyingExpressions,
            studyingWords: VocabStore.shared.studying,
            sessionId: sessionId, sessionStartedAt: session.startedAt)
        computed.carryovers = carryovers
        report { $0.carryovers = carryovers.count }

        // Free-talk sessions (no picked topic) take the summary's generated
        // title so History/Practice lists don't fill with identical
        // "Conversation" rows.
        let generatedTitle = payload.title?.trimmingCharacters(in: .whitespacesAndNewlines)
        let existingTopic = session.topic?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let resolvedTopic = existingTopic.isEmpty ? (generatedTitle ?? "") : existingTopic

        var saved = session
        saved.topic = resolvedTopic.isEmpty ? nil : resolvedTopic
        saved.summary = computed
        SessionStore.shared.save(saved)

        // Clear cards from a previous analysis of THIS session (resume
        // re-summarizes the whole thing, and so does a regenerate) so they
        // don't pile up — but ONLY the untouched ones. A card the learner
        // has already reviewed carries Leitner progress a re-analysis must
        // not reset; ingest's dedupe skips re-minting the survivors.
        DrillStore.shared.clearUnreviewedCards(for: sessionId)
        let mintedCards = DrillStore.shared.ingest(summary: computed, turns: turns,
                                                   sessionId: sessionId)
        report { $0.cards = mintedCards }
        // Producing a card's phrase live outranks any flashcard tap — credit
        // it against the SRS schedule, not just the wrap-up.
        DrillStore.shared.markUsedInConversation(
            ids: carryovers.filter { $0.source == .drillCard }.compactMap { $0.sourceId })
        // Same principle for book material: producing it live masters it,
        // wherever the book lives.
        appState.markCurriculumItemsUsedInConversation(
            itemIds: carryovers.filter { $0.source == .curriculumItem }.compactMap { $0.sourceId })
        if !carryovers.isEmpty {
            Analytics.capture("carryovers_detected", [
                "count": carryovers.count,
                "from_cards": carryovers.filter { $0.source == .drillCard }.count,
                "from_suggestions": carryovers.filter { $0.source == .suggestion }.count,
            ])
        }
        // Grow the long-term learner profile — the next conversation's system
        // prompt picks these patterns up.
        appState.recordSessionOutcome(summary: computed, turns: turns)
        // The other memory: what the talk taught the fluent self about the
        // PERSON. A learner tells their future self about their job once, out
        // loud, and expects it to be known next week — so it's written into
        // the same profile they fill in by hand and read back through
        // `personaBlock`. Capped at 3 a session: this is a notebook, not a
        // transcript, and the model volunteers more than it should when a talk
        // ran long.
        let learned = payload.about_user
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty && $0.count <= 140 }
            .prefix(3)
            .map { PersonaNote(text: $0, sessionId: sessionId, learnedAt: Date()) }
        // A plain free talk is where the fluent self gets to know someone —
        // a scenario casts it as a barista and a Find-people call as a
        // stranger, and neither of those met the learner. Stamping this is
        // what retires the first-call framing.
        let wasIntroTalk = session.counterpartId == nil && existingTopic.isEmpty
        if !learned.isEmpty || (wasIntroTalk && appState.persona?.metAt == nil) {
            appState.rememberAboutUser(
                Array(learned),
                metAt: wasIntroTalk ? (session.endedAt ?? Date()) : nil)
        }
        // Kick off async weekly-report generation if unlock conditions are
        // met. Fires-and-forgets — UI doesn't block on Gemini.
        appState.maybeGenerateWeeklyReport()
        // Fresh cards just landed in the queue — (re)schedule the due
        // reminder. This is the one contextual moment where asking for
        // notification permission makes sense.
        Task { await DrillReminder.reschedule(allowPermissionPrompt: true) }
        // Write tomorrow's call NOW, off the talk that just ended.
        //
        // Not in the morning: iOS won't reliably run background work at a
        // chosen hour, and a call that fails to generate is a call that never
        // rings. Here the app is foreground, the network is up, and the
        // context is as fresh as it will ever be — this session's own topic
        // and phrases. By 8am the script and its audio are already on disk,
        // so the ring works offline and cannot fail.
        appState.refreshDailyCall(force: true)

        report { $0.finished = true }
        return Result(summary: computed, session: saved)
    }
}

File: android/app/src/main/java/com/roro/futurevoice/talk/SessionSummarizer.kt
package com.roro.futurevoice.talk

import android.content.Context
import android.util.Log
import com.roro.futurevoice.data.AuthRepository
import com.roro.futurevoice.data.CefrLevel
import com.roro.futurevoice.data.DrillIngest
import com.roro.futurevoice.data.DrillStore
import com.roro.futurevoice.data.ProfileStore
import com.roro.futurevoice.data.SessionStore
import com.roro.futurevoice.data.StoreEvents
import com.roro.futurevoice.data.StoreJson
import com.roro.futurevoice.data.VocabStore
import com.roro.futurevoice.net.EdgeError
import com.roro.futurevoice.net.SessionSummaryClient
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/**
 * Turning a finished talk into its review material — the Android twin of
 * `SessionSummarizer.swift`, in stages. What is here now: the model call
 * (server-side prompt), a LENIENT decode (an element the model got wrong is
 * skipped, never the whole list), the two verbatim guards (expressions the
 * learner "used" must appear in their own turns, "offered" ones in the
 * fluent self's and not the learner's; grammar quotes likewise, and a fix
 * that only changes punctuation/casing is dropped), the generated title
 * becoming a free talk's topic, and the save. Still to port, with vectors:
 * VocabStore ingest, DrillStore minting, CarryoverDetector, LearnerProfile
 * absorb, persona notes (`about_user` is decoded and kept on the result for
 * that day).
 *
 * Runs on an app-lifetime scope: the call screen is gone by the time the
 * model answers.
 */
object SessionSummarizer {

    private const val TAG = "SessionSummarizer"
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    /** Session ids being analyzed right now — what a "Working…" button reads. */
    private val _inFlight = MutableStateFlow<Set<String>>(emptySet())
    val inFlight: StateFlow<Set<String>> = _inFlight

    data class Progress(
        val readBack: Boolean = false,
        val wroteCorrections: Boolean = false,
        val wroteDrills: Boolean = false,
        val wroteExpressions: Boolean = false,
        val wroteGrammar: Boolean = false,
        val finished: Boolean = false,
    ) {
        /** Same key-order reading as `Progress.absorb(partial:)` on iOS. */
        fun absorb(partial: String) = copy(
            readBack = readBack || partial.contains("\"phrases_used\""),
            wroteCorrections = wroteCorrections || partial.contains("\"new_patterns_detected\""),
            wroteDrills = wroteDrills || partial.contains("\"expressions_used\""),
            wroteExpressions = wroteExpressions || partial.contains("\"weak_vocab_areas\""),
            wroteGrammar = wroteGrammar || partial.contains("\"overall_note\""),
        )
    }

    /** What a talk is allowed to yield follows how much was said in it. */
    fun expressionBudget(fluentTurns: Int): Int = minOf(14, maxOf(6, fluentTurns))

    fun needsSummary(session: Session): Boolean =
        session.summary == null && session.turns.any { it.role == TurnRole.USER }

    /** Fire-and-forget from the call screen; the store bumps when done. */
    fun summarizeInBackground(context: Context, session: Session, nativeLanguage: String, level: CefrLevel) {
        if (!needsSummary(session)) return
        if (session.id in _inFlight.value) return
        val appContext = context.applicationContext
        _inFlight.update { it + session.id }
        scope.launch {
            try {
                runCatching { summarize(appContext, session, nativeLanguage, level) }
                    .onFailure { Log.w(TAG, "summary failed for ${session.id}: ${it.message}") }
            } finally {
                _inFlight.update { it - session.id }
            }
        }
    }

    suspend fun summarize(
        context: Context,
        session: Session,
        nativeLanguage: String,
        level: CefrLevel,
        onProgress: ((Progress) -> Unit)? = null,
    ): Session {
        val turns = session.turns
        val auth = AuthRepository()
        val client = SessionSummaryClient(auth)
        val metrics = ScorecardMetrics.compute(turns)
        var progress = Progress()

        // The REAL learner profile (what past sessions taught) — read before
        // the call, absorbed after it, exactly iOS's order.
        val profileStore = ProfileStore.shared(context)
        val profile = profileStore.load(session.targetLanguage, level.code)
        val body = SessionSummaryClient.RequestBody(
            target_language = session.targetLanguage,
            native_language = nativeLanguage,
            profile = Json.parseToJsonElement(
                StoreJson.json.encodeToString(LearnerProfile.serializer(), profile)).jsonObject,
            known_about_user = emptyList(),
            expression_budget = expressionBudget(turns.count { it.role == TurnRole.FLUENT_SELF }),
            transcript = formatTranscript(turns),
            metrics = metrics.promptJson(),
        )
        // Stable key: a retry re-runs the SAME logical request for free as
        // long as the transcript hasn't grown (iOS: "summary:<id>:<turns>").
        val key = "summary:${session.id}:${turns.size}"
        // The model occasionally writes a shape we can't read — most often a
        // native-language note carrying unescaped ASCII quotes (「"going"」).
        // The learner did nothing wrong: ask ONCE more before making the end
        // of a talk look like a failure. Free — the ledger dedupes on the key.
        // (iOS: the `isMalformedModelOutput` retry in SessionSummarizer.)
        fun parse(text: String): JsonObject = Json.parseToJsonElement(text).jsonObject
        val payload = try {
            parse(client.summarize(body, key) { partial ->
                progress = progress.absorb(partial); onProgress?.invoke(progress)
            })
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.w(TAG, "summary payload unreadable (${e.message?.take(80)}) — retrying once")
            parse(client.summarize(body, key))
        }
        var computed = toDomain(payload)

        // Verbatim guards — the same normalization CarryoverDetector uses.
        val userTexts = turns.filter { it.role == TurnRole.USER }.map { it.transcript }
        val haystack = normalized(userTexts.joinToString(" "))
        val verifiedUsed = computed.expressionsUsed.filter { p ->
            val n = normalized(p); n.isNotEmpty() && haystack.contains(n)
        }
        val fluentHaystack = normalized(
            turns.filter { it.role == TurnRole.FLUENT_SELF }.joinToString(" ") { it.transcript })
        val alreadyMine = verifiedUsed.map(::normalized).toSet()
        val seenOffered = HashSet<String>()
        val verifiedOffered = computed.expressionsOffered.filter { p ->
            val n = normalized(p)
            n.isNotEmpty() && n !in alreadyMine && !haystack.contains(n) &&
                fluentHaystack.contains(n) && seenOffered.add(n)
        }
        val priorUsed = session.summary?.expressionsUsed.orEmpty()
        val priorOffered = session.summary?.expressionsOffered.orEmpty()
        val grammar = computed.grammarIssues.filter { g ->
            val needle = normalized(g.quote)
            needle.isNotEmpty() && haystack.contains(needle) && needle != normalized(g.correction)
        }
        computed = computed.copy(
            expressionsUsed = priorUsed + verifiedUsed.filter { it !in priorUsed },
            expressionsOffered = priorOffered + verifiedOffered.filter { it !in priorOffered },
            grammarIssues = grammar,
        )

        // ── The loop closes (`SessionSummarizer.swift`, in stages) ──
        val language = session.targetLanguage
        val vocab = VocabStore.shared(context)
        val drills = DrillStore.shared(context)

        // Words into the long-term pool; merge with the previous summary's
        // list so a resumed talk can't erase "words you used first".
        val freshWords = vocab.ingest(session.id, userTexts, language)
        val priorWords = session.summary?.newWordsUsed.orEmpty()
        // Expressions: seed pre-tracking sessions, then count this batch.
        vocab.notePriorExpressions(session.id, priorUsed, language)
        vocab.ingestExpressions(session.id, verifiedUsed, language)

        // Carryovers run against the cards as they stood BEFORE this
        // session's own corrections are ingested below.
        val carryovers = CarryoverDetector.detect(
            turns = turns,
            cards = drills.load(language),
            studyingExpressions = vocab.studyingExpressions(language),
            studyingWords = vocab.studying(language),
            sessionId = session.id, sessionStartedAt = session.startedAt,
            language = language,
        )
        computed = computed.copy(
            newWordsUsed = priorWords + freshWords.filter { it !in priorWords },
            carryovers = carryovers,
        )

        // A free talk takes the generated title so lists don't fill with "Conversation".
        val generatedTitle = payload["title"]?.jsonPrimitive?.contentOrNull?.trim().orEmpty()
        val existingTopic = session.topic?.trim().orEmpty()
        val resolvedTopic = existingTopic.ifEmpty { generatedTitle }.ifEmpty { null }

        val saved = session.copy(topic = resolvedTopic, summary = computed)
        SessionStore.shared(context).save(saved)

        // Cards: clear this session's untouched ones (a re-analysis must not
        // reset Leitner progress), mint, then credit live production.
        drills.clearUnreviewedCards(session.id, language)
        val minted = DrillIngest.mint(drills.load(language), computed, turns, session.id,
            System.currentTimeMillis())
        drills.upsertMany(minted, language)
        drills.markUsedInConversation(
            carryovers.filter { it.source == Carryover.Source.DRILL_CARD }.mapNotNull { it.sourceId },
            language)

        // Grow the long-term profile — the next conversation's prompt reads it.
        val speakingSeconds = turns.filter { it.role == TurnRole.USER }
            .sumOf { it.durationMs / 1000.0 }
        profile.absorb(computed, speakingSeconds)
        profileStore.save(profile)

        StoreEvents.bump()
        progress = progress.copy(finished = true); onProgress?.invoke(progress)
        return saved
    }

    /** `ConversationEngine.formatTranscript` — role-labelled lines. */
    fun formatTranscript(turns: List<Turn>): String =
        turns.joinToString("\n") { t ->
            "[${if (t.role == TurnRole.USER) "USER" else "FLUENT_SELF"}] ${t.transcript}"
        }

    /** `CarryoverDetector.normalized`: letters/digits/spaces only, lowercase, single-spaced. */
    fun normalized(text: String): String =
        text.filter { it.isLetterOrDigit() || it.isWhitespace() }
            .lowercase().split(Regex("\\s+")).filter { it.isNotEmpty() }.joinToString(" ")

    // ── Lenient decode (`ClaudeSummaryPayload.lossyArray` + `toDomain`) ──

    private fun str(o: JsonObject, k: String): String? =
        (o[k] as? JsonElement)?.let { runCatching { it.jsonPrimitive.contentOrNull }.getOrNull() }

    private fun strings(o: JsonObject, k: String): List<String> =
        (o[k] as? JsonArray)?.mapNotNull { runCatching { it.jsonPrimitive.contentOrNull }.getOrNull() }.orEmpty()

    private fun objects(o: JsonObject, k: String): List<JsonObject> =
        (o[k] as? JsonArray)?.mapNotNull { it as? JsonObject }.orEmpty()

    private fun toDomain(p: JsonObject): SessionSummary {
        val now = System.currentTimeMillis()
        val scorecard = (p["scorecard"] as? JsonObject)?.let { sc ->
            fun axis(k: String): AxisScore? {
                val a = sc[k] as? JsonObject ?: return null
                val score = a["score"]?.jsonPrimitive?.intOrNull ?: return null
                return AxisScore(score.coerceIn(0, 100), str(a, "note").orEmpty())
            }
            val v = axis("vocabulary"); val g = axis("grammar"); val e = axis("expressiveness")
            val fRaw = (sc["fluency"] as? JsonObject)?.get("score")?.jsonPrimitive?.intOrNull
            if (v == null || g == null || e == null || fRaw == null) null
            else SessionScorecard(
                vocabulary = v, grammar = g, expressiveness = e,
                // -1 signals "no timing data" — clamp and say so, as iOS does.
                fluency = if (fRaw < 0) AxisScore(0, "Speak a bit more next session for a fluency read.")
                else axis("fluency")!!,
                topLine = str(sc, "top_line").orEmpty(),
                cefrLevel = str(sc, "cefr_level")?.lowercase(),
            )
        }
        return SessionSummary(
            phrasesUsed = objects(p, "phrases_used").mapNotNull { o ->
                val said = str(o, "user_said") ?: return@mapNotNull null
                val alt = str(o, "fluent_alternative") ?: return@mapNotNull null
                PhraseFeedback(userSaid = said, fluentAlternative = alt, reason = str(o, "reason").orEmpty())
            },
            newPatternsDetected = objects(p, "new_patterns_detected").mapNotNull { o ->
                val m = str(o, "mistake") ?: return@mapNotNull null
                val c = str(o, "correction") ?: return@mapNotNull null
                val freq = when (str(o, "frequency_hint")?.lowercase()) { "often" -> 5; "sometimes" -> 2; else -> 1 }
                LearnerPattern(mistake = m, correction = c, context = str(o, "context").orEmpty(),
                    frequency = freq, lastSeenAt = now)
            },
            suggestedDrills = strings(p, "suggested_drills"),
            overallNote = str(p, "overall_note").orEmpty(),
            scorecard = scorecard,
            expressionsUsed = strings(p, "expressions_used"),
            expressionsOffered = strings(p, "expressions_offered"),
            weakVocabAreas = strings(p, "weak_vocab_areas"),
            grammarIssues = objects(p, "grammar_errors").mapNotNull { o ->
                val q = str(o, "quote")?.trim().orEmpty(); val c = str(o, "correction")?.trim().orEmpty()
                if (q.isEmpty() || c.isEmpty()) null else GrammarIssue(quote = q, correction = c, note = str(o, "note").orEmpty())
            },
        )
    }
}
